src/config.rs

```diff
@@ -90,6 +65,156 @@ impl Config {
         }
         Ok(())
     }
+
+    fn load_project_config() -> Result<Self> {
+        let default_path = env::HK_FILE
+            .as_ref()
+            .map(|s| s.as_str())
+            .unwrap_or("hk.pkl");
+        let paths = vec![default_path, "hk.toml", "hk.yaml", "hk.yml", "hk.json"];
+        let mut cwd = std::env::current_dir()?;
+        while cwd != Path::new("/") {
+            for path in &paths {
+                let path = cwd.join(path);
+                if path.exists() {
+                    let hash_key = format!("{}.json", hash::hash_to_str(&path));
+                    let hash_key_path = env::HK_CACHE_DIR.join("configs").join(hash_key);
+                    return CacheManagerBuilder::new(hash_key_path)
+                        .with_fresh_file(path.to_path_buf())
+                        .build()
+                        .get_or_try_init(|| {
+                            Self::read(&path).wrap_err_with(|| {
+                                format!("Failed to read config file: {}", path.display())
+                            })
+                        })
+                        .cloned();
+                }
+            }
+            cwd = cwd.parent().map(PathBuf::from).unwrap_or_default();
+        }
+        debug!("No config file found, using default");
+        let mut config = Config::default();
+        config.init(Path::new(default_path))?;
+        Ok(config)
+    }
+
+    fn apply_user_config(&mut self, user_config: &Option<UserConfig>) -> Result<()> {
+        if let Some(user_config) = user_config {
```

Comment:
this is way beyond the scope of this PR but at some point I think it'll be important to do something for this configuration like I did in mise where all the settings are codegen'ed from a file. In mise this is https://github.com/jdx/mise/blob/main/settings.toml

there's just a lot of boilerplate we'll need to add around new config